consensus/event_loop.go

```diff
@@ -393,14 +395,125 @@ func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
+			EventType:   ConsensusEventTypeConstructVoteQC, // The event type
+			View:        fc.currentView,                    // The view that we can construct a block on
+			BlockHash:   fc.getTipBlock().GetBlockHash(),   // The block hash that we can build the QC for
+			BlockHeight: fc.getTipBlock().GetHeight() + 1,  // The block height that we can build the QC for
+			QC: &quorumCertificate{
+				blockHash:           fc.getTipBlock().GetBlockHash(), // The previous block's block hash
+				view:                fc.getTipBlock().GetView(),      // The previous block's view
+				signersList:         signersList,                     // The signers list who voted on the previous block
+				aggregatedSignature: aggregatedSignature,             // Aggregated signature from votes on the previous block
+			},
+		}
+
+		// Cancel the block construction task since we know we can construct a QC in the current view.
+		// It will be rescheduled when we advance view.
+		fc.nextBlockConstructionTask.Cancel()
+
+		return
+	}
+
+	// TODO: Check if we can construct a timeout QC for the current view, and send the signal to the
+	// server
+
+	return
+}
+
+func (fc *FastHotStuffEventLoop) tryConstructQCInCurrentView() (
+	_success bool, // true if and only if we are able to construct a QC for the tip block in the current view
+	_signersList *bitset.Bitset, // bitset of signers for the aggregated signature for the tip block
+	_aggregateSignature *bls.Signature, // aggregated signature for the tip block
+) {
+	// If currentView != tipBlock.View + 1, then we have timed out at some point, and can no longer
+	// construct a block with a QC of votes from the chain tip.
+	tipBlock := fc.getTipBlock()
+	if fc.currentView != tipBlock.GetView()+1 {
+		return false, nil, nil
+	}
+
+	// Fetch the validator set at the tip block
+	validatorSet := fc.getTipValidatorSet()
+
+	// Compute the chain tip's signature payload.
+	voteSignaturePayload := GetVoteSignaturePayload(tipBlock.GetView(), tipBlock.GetBlockHash())
+
+	// Fetch the validator votes for the chain tip
+	votesByValidator := fc.votesSeen[voteSignaturePayload]
+
+	// Compute the total stake and total stake with votes
+	totalStake := uint256.NewInt()
+	totalStakeWithVotes := uint256.NewInt()
+
+	// Track the signatures and signers list for the chain tip
+	signersList := bitset.NewBitset()
+	signatures := []*bls.Signature{}
+
+	// Iterate through the entire validator set and check if each one has voted for the tip block. Track
+	// all voters and their stakes.
+	for ii, validator := range validatorSet {
+		totalStake = uint256.NewInt().Add(totalStake, validator.GetStakeAmount())
+
+		// Skip the validator if it hasn't voted for the the block
+		vote, hasVoted := votesByValidator[validator.GetPublicKey().ToString()]
+		if !hasVoted {
+			continue
+		}
+
+		// Verify the vote signature
+		if !isValidSignatureSinglePublicKey(vote.GetPublicKey(), vote.GetSignature(), voteSignaturePayload[:]) {
```

Comment:
Would it be more efficient for us to check multiple pub keys against the payload in one go? Just something to noodle on - not sure how much more efficient it is to use `flowCrypto.VerifyBLSSignatureOneMessage` with many pub keys vs. many calls to `flowCrypto.VerifyBLSSignatureOneMessage` with a single pub key.

Reply:
The current code is O(n) with respect to the number of votes. Every public key has a separate signature in its vote messages. Note, the signatures aren't aggregated. In order to perform a multi-pubkey signature verification, we would need to first aggregate the signatures which would itself be an O(n) operation.

Reply:
yeah this is not a today problem.